### DISCUSSION TITLE
Exclude Classes from Average (e.g. macro f1-score)

### DISCUSSION THREAD

**Discussion by plonerma** [Lightning-AI/torchmetrics#1721] [Classification]
I would like to exclude certain classes from the averaging in the classification metrics (accuracy, f1, recall, precision, etc). However, I would want to still include classes in the prediction (e.g. in an NER problem, a tagger might predict that a token is not part of a named entity and therefor predict an "out" or "O"). As they still affect the false positives and negatives, `ignore_index` does not work in this case. What would be the best way to solve this using the torchmetrics framework?

As far as I can see, the only option at the moment would be to set average to none and compute the desired averages post-hoc. However, this would require writing a wrapper for all of the used classification metrics (instead of addressing the underlying StatScores)

I would appreciate any pointer in how to solve this, or how to best address this in a PR if that is a feature which might be interesting in other use-cases as well.

**Comment by plonerma** [comment]
I opened a [feature-request](https://github.com/Lightning-AI/torchmetrics/issues/1723) in the issues regarding this.

**Comment by Borda** [comment]
This is a legit use case (NER with "O" tag, segmentation with "void" class) and there's still no built-in `exclude_classes` parameter. Here's the cleanest pattern today (v1.9.0):

**Option 1 — `average="none"` + manual selection:**
```python
from torchmetrics.classification import MulticlassF1Score
import torch

num_classes = 10
exclude_classes = {0, 7}  # e.g., "O" tag and padding
include_mask = torch.tensor([i not in exclude_classes for i in range(num_classes)])

f1 = MulticlassF1Score(num_classes=num_classes, average="none")
per_class = f1(preds, target)  # shape (num_classes,)
macro_f1_filtered = per_class[include_mask].mean()
```

**Option 2 — Wrap it in a custom metric for Lightning:**
```python
from torchmetrics import Metric
from torchmetrics.classification import MulticlassF1Score


class FilteredMacroF1(Metric):
    is_differentiable = False
    higher_is_better = True

    def __init__(self, num_classes: int, exclude_classes: list[int], **kwargs):
        super().__init__(**kwargs)
        self.f1 = MulticlassF1Score(num_classes=num_classes, average="none", **kwargs)
        mask = torch.ones(num_classes, dtype=torch.bool)
        mask[exclude_classes] = False
        self.register_buffer("include_mask", mask)

    def update(self, preds, target):
        self.f1.update(preds, target)

    def compute(self):
        per_class = self.f1.compute()
        return per_class[self.include_mask].mean()

    def reset(self):
        super().reset()
        self.f1.reset()
```

Note: `ignore_index` is NOT what you want here — it removes samples with that target entirely. What you want is to compute all class-level metrics but exclude certain classes from the macro average.

Docs: [MulticlassF1Score](https://lightning.ai/docs/torchmetrics/stable/classification/f1_score.html)
